File: 1_Introduction/py_code/clock_speed.py
import time

GIGA = 1000000000
ITERS = 1000000


def main():
    print("SUM")
    measure_loop(sum_loop)
    print("MULT")
    measure_loop(mult_loop)
    print("DIV")
    measure_loop(div_loop)


def sum_loop():
    sum = 0
    for i in range(ITERS):
        sum += i


def mult_loop():
    sum = 1
    for i in range(1, ITERS):
        sum *= i


def div_loop():
    sum = 1
    for i in range(1, ITERS):
        sum /= i


def measure_loop(func):
    start = time.process_time()
    func()
    end = time.process_time()
    ops = ITERS / (end - start)
    print("Clock speed approx {:.3f} GHz".format(ops / GIGA))

    s = time.time()
    func()
    e = time.time()
    ops = ITERS / (e - s)
    print("time.time speed {:.3f} GHz".format(ops / GIGA))


main()

File: 1_Introduction/go_code/clockSpeed.go
package main

import (
	"fmt"
	"time"
)

const ITERS = 1000000000

type f func()

func sumLoop() {
	sum := 0
	for i := 0; i < ITERS; i++ {
		sum += i
	}
}

func multLoop() {
	sum := 1
	for i := 1; i < ITERS; i++ {
		sum *= i
	}
}

func divLoop() {
	sum := 1
	for i := 1; i < ITERS; i++ {
		sum /= i
	}
	// fmt.Printf("%d", sum)
}

func printSpeed(fn f) {
	start := time.Now()
	fn()
	end := time.Now()
	elapsed := end.Sub(start)
	ops := float64(ITERS) / elapsed.Seconds()
	fmt.Printf("Clock speed approx %.3f GHz\n", ops/float64(ITERS))

}

func main() {
	fmt.Printf("SUM\n")
	printSpeed(sumLoop)
	fmt.Printf("MULT\n")
	printSpeed(multLoop)
	fmt.Printf("DIV\n")
	printSpeed(divLoop)
}
